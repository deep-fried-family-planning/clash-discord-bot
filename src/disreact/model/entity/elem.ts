import {Prim} from '#src/disreact/model/entity/elem/prim.ts'
import {Rest} from '#src/disreact/model/entity/elem/rest.ts'
import {Task} from '#src/disreact/model/entity/elem/task.ts'
import {FC} from '#src/disreact/model/entity/fc.ts'

export * from '#src/disreact/model/entity/elem/frag.ts'
export * from '#src/disreact/model/entity/elem/prim.ts'
export * from '#src/disreact/model/entity/elem/rest.ts'
export * from '#src/disreact/model/entity/elem/task.ts'

export * as Elem from '#src/disreact/model/entity/elem.ts'
export type Elem =
  | Rest
  | Task

export type Any =
  | Elem
  | Prim

export interface MetaProps {
  type   : any
  id?    : string | undefined
  ids?   : string | undefined
  idn?   : string | undefined
  idx?   : number | undefined
  parent?: Elem | undefined
  props  : any
  nodes  : Any[]
}

export const jsx = (type: any, props: any): Elem => {
  if (type === undefined) {
    return props.children
  }

  switch (typeof type) {
    case 'string':
      return Rest.jsxRest(type, props)
    case 'function':
      return Task.jsxTask(type, props)
    default:
      throw new Error()
  }
}

export const jsxs = (type: any, props: any): Elem => {
  props.children = props.children.flat()

  if (type === undefined) {
    return props.children
  }

  switch (typeof type) {
    case 'string':
      return Rest.jsxsRest(type, props)
    case 'function':
      return Task.jsxsTask(type, props)
    default:
      throw new Error()
  }
}

export const jsxDEV = (type: any, props: any): Elem => {
  if (!Array.isArray(props.children)) {
    return jsx(type, props)
  }
  return jsxs(type, props)
}

export const cloneElem = (self: Elem) => {
  if (Rest.isRest(self)) {
    return Rest.cloneRest(self)
  }

  return Task.cloneTask(self)
}

export const deepCloneElem = <A extends Any>(self: A): A => {
  if (Prim.isPrim(self)) {
    return Prim.clonePrim(self) as A
  }

  const cloned = cloneElem(self)

  for (let i = 0; i < cloned.nodes.length; i++) {
    const node = cloned.nodes[i]
    cloned.nodes[i] = deepCloneElem(node)
  }

  return cloned as A
}

export const linearizeElem = <A extends Elem>(self: A): A => {
  if (Task.isTask(self)) {
    delete self.strand.nexus
    delete self.strand.elem
  }

  for (let i = 0; i < self.nodes.length; i++) {
    const node = self.nodes[i]

    if (!Prim.isPrim(node)) {
      linearizeElem(node)
    }
  }

  return self
}

export const findElem = (self: Elem, predicate: (elem: Elem) => boolean): Elem | false => {
  if (predicate(self)) {
    return self
  }

  for (let i = 0; i < self.nodes.length; i++) {
    const node = self.nodes[i]

    if (Prim.isPrim(node)) {
      continue
    }

    const found = findElem(node, predicate)

    if (found) {
      return found
    }
  }
  return false
}

export const isSameElem = (a: Elem, b: Elem): boolean => {
  if (a === b) {
    return true
  }
  if (a.type !== b.type) {
    return false
  }
  if (a.constructor.name !== b.constructor.name) {
    return false
  }
  return true
}

export const setIds = (children: Any[], parent: Elem) => {
  for (let i = 0; i < children.length; i++) {
    const child = children[i]

    if (Prim.isPrim(child)) {
      continue
    }

    if (Rest.isRest(child)) {
      child.ids = `${child.type}:${i}`
    }

    if (Task.isTask(child)) {
      child.ids = `${FC.getName(child.type)}:${i}`
    }
  }

  return children
}

export const connectChild = (parent: Elem, child: Elem, idx: number) => {
  child.parent = parent
  child.idx = idx

  const child_id = `${child.idn ?? child.type}:${idx}`
  const parent_id = `${parent.idn ?? parent.type}:${parent.idx}`

  child.id = `${parent.id}:${child_id}`
  child.ids = `${parent_id}:${child_id}`
}
