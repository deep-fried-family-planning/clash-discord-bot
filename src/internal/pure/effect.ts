import {Order} from 'effect';


export {
    Effect as E,
    Console as CSL,
    Data as D,
    DateTime as DT,
    Redacted as RDT,
    Config as CFG,
    Context as CTX,
    Layer as L,
    Cache as C,
    Logger as Logger,
    LogLevel,
    LogSpan,
    Schema as S,
    Brand as B,
    RateLimiter as RL,
    Metric as Metric,
    Order as ORD,
    Option as O,
    Equal as EQ,

    Cron,

    pipe,
    pipe as p,
    flow,
    flow as f,

} from 'effect';


export {
    gen as g,

} from 'effect/Effect';


export const ORDS = Order.string;
export const ORDSR = Order.reverse(Order.string);
export const ORDN = Order.number;
export const ORDNR = Order.reverse(Order.number);


export * as Ar from 'effect/Array'; // collection array
export * as Kv from 'effect/Record'; // collection key-value
