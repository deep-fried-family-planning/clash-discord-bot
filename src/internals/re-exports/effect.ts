export {
    Effect as E,

    Config as Cfg,
    Redacted as Rdt,

    Console as CSL,

    Data as D,
    DateTime as DT,

    Redacted as RDT,
    Config as CFG,
    Context as CTX,
    Layer as L,

    Cache as C,

    Logger as Logger,
    LogLevel,
    LogSpan,

    Schema as S,

    Brand as B,

    RateLimiter as RL,

    Metric as Metric,

    flow,
    pipe,

} from 'effect';
