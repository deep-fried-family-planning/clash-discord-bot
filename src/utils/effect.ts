export {
    Effect as E,

    Config as Cfg,
    Redacted as Rdt,

    Layer as L,

    Logger as Logger,
    LogLevel,
    LogSpan,

    RateLimiter as RL,

    Metric as Metric,

    flow,
    pipe,

} from 'effect';
