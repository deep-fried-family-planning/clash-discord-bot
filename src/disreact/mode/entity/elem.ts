import type * as Event from '#src/disreact/mode/entity/event.ts';
import * as FC from '#src/disreact/mode/entity/fc.ts';
import * as Props from '#src/disreact/mode/entity/props.ts';
import * as Array from 'effect/Array';
import * as Data from 'effect/Data';

export declare namespace Elem {
  export type Val = {
    type : false;
    props: any;
    nodes: Any[];
  };
  export type Api = {
    id?     : string;
    type    : string;
    props   : Props.Props;
    nodes   : Any[];
    handler?: Event.Handler | undefined;
  };
  export type Fn = {
    id?  : string;
    type : FC.FC.Any<any, any>;
    props: Props.Props;
    nodes: Any[];
  };
  export type Node = Api | Fn;
  export type Any = Val | Node;
  export type Children = Any | Children[];
  export type Elem = Any;
}
export type Elem = Elem.Any;
export type Val = Elem.Val;
export type Api = Elem.Api;
export type Fn = Elem.Fn;
export type Node = Elem.Node;
export type Any = Elem.Any;
export type Children = Elem.Children;

const emptyNodes = () => Data.array([] as Elem.Any[]) as Elem.Any[];

export const ensureChildren = (cs?: Elem.Any | Elem.Any[]): Elem.Any[] => {
  if (!cs) return emptyNodes();
  if (Array.isArray(cs)) return Data.array(cs.flat()) as Elem.Any[];
  return Data.array([cs]) as Elem.Any[];
};

export const isVal = (e: Elem.Any): e is Elem.Val => !e.type;
export const isApi = (e: Elem.Any): e is Elem.Api => typeof e.type === 'string';
export const isFn = (e: Elem.Any): e is Elem.Fn => FC.isFC(e.type);

export const val = (value: any): Elem.Val => {
  return Data.struct({
    type : false,
    props: value,
    nodes: emptyNodes(),
  });
};

export const api = (tag: string, props: Props.Props): Elem.Api => {
  const handler = Props.extractHandler(props);
  return Data.struct({
    type : tag,
    props: Props.make(props),
    nodes: ensureChildren(props.children),
    handler,
  });
};

export const fn = (tag: any, props: Props.Props): Elem.Fn => {
  return Data.struct({
    type : FC.make(tag),
    props: Props.make(props),
    nodes: emptyNodes(),
  });
};
