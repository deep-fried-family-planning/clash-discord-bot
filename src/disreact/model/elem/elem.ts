import {Props} from '#src/disreact/model/elem/props.ts';
import type {Trigger} from '#src/disreact/model/elem/trigger.ts';
import {FC} from '#src/disreact/model/elem/fc.ts';
import {Fibril} from '#src/disreact/model/meta/fibril.ts';
import {E} from '#src/disreact/utils/re-exports.ts';

export * as Elem from '#src/disreact/model/elem/elem.ts';
export type Elem =
  | Task
  | Rest
  | Fragment
  | Value;

export interface Node {
  type   : any;
  id?    : string | undefined;
  ids?   : string | undefined;
  idn?   : string | undefined;
  idx?   : number | undefined;
  length?: number | undefined;
  fibril?: Fibril | undefined;
  props  : Props;
  nodes  : Elem[];
}

export const isNode = (self: any): self is Node => typeof self === 'object';

export type Child =
  | Task
  | Rest
  | Fragment
  | ValueType;

export type Children = Child | Child[];

export const isChild = (children: any): children is Child => !children.length;

export const connectChildren = (parent: Node, children: Children): Elem[] => {
  if (isValueType(children)) return [makeValue(children)];
  if (isChild(children)) return [connectChild(parent, children)];
  return connectNodes(parent, children);
};

export const connectNodes = (parent: Node, children: Child[]): Elem[] => {
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    children[i] = isValueType(child)
      ? makeValue(child)
      : connectChild(parent, child, i);
  }
  return children as Elem[];
};

export const connectChild = (parent: Node, child: Node, idx: number = 0): Elem => {
  child.idx = idx;
  const child_id = `${child.idn ?? child.type ?? ''}:${idx}`;
  const parent_id = `${parent.idn ?? parent.type ?? ''}:${parent.idx}`;
  child.id = `${parent.id}:${child_id}`;
  child.ids = `${parent_id}:${child_id}`;
  return child;
};

export type TaskType<E = any, R = any> = (prop?: any) =>
  | Child
  | Promise<Child>
  | E.Effect<Child, E, R>;

export interface Task extends Node {
  type  : FC;
  fibril: Fibril;
}

export const isTaskType = (self: any): self is TaskType => typeof self === 'function';

export const isTask = (self: any): self is Task => typeof self === 'object' && typeof self.type === 'function';

export const isEqualTask = (a: Task, b: Task) => {
  if (a === b) return true;
  if (a.type !== b.type) return false;
  if (!Props.isEqual(a.props, b.props)) return false;
  if (a.nodes.length !== b.nodes.length) return false;
  return true;
};

export const makeTask = (type: any, props: any): Task => {
  const fc = FC.make(type);
  return {
    type  : fc,
    idn   : FC.getName(fc),
    props,
    nodes : [],
    fibril: Fibril.make(),
  };
};

export const cloneTask = (self: Task): Task => {
  const {type, props, nodes, fibril, ...rest} = self;
  const cloned = structuredClone(rest) as Task;
  cloned.type = type;
  cloned.fibril = Fibril.clone(fibril);
  cloned.props = Props.deepCloneTaskProps(props);
  cloned.nodes = nodes;
  return cloned;
};

export type RestType = string;

export interface Rest extends Node {
  type    : string;
  handler?: Trigger.Handler<any> | undefined;
}

export const isRestType = (self: any): self is RestType => typeof self === 'string';

export const isRest = (self: any): self is Rest => typeof self === 'object' && typeof self.type === 'string';

export const makeRest = (type: string, props: any, nodes: any[]): Rest =>
  ({
    type,
    props,
    nodes,
    handler: Props.getHandler(props),
  });

export const cloneRest = (self: Rest): Rest => {
  const {props, nodes, handler, ...rest} = self;
  const cloned = structuredClone(rest) as Rest;
  cloned.props = Props.cloneKnownProps(props);
  cloned.nodes = nodes;
  cloned.handler = handler;
  return cloned;
};

/**
 * Fragment
 */
export type FragmentType = undefined;

export const FragmentType = undefined;

export interface Fragment extends Node {
  type: FragmentType;
};

export const isFragmentType = (self: any): self is FragmentType => self === FragmentType;

export const isFragment = (self: any): self is Fragment => self.type === FragmentType;

export const makeFragment = (type: undefined, props: any): Fragment =>
  ({
    type,
    props,
    nodes: props.children,
  });

export const cloneFragment = (self: Fragment) => {
  const {props, nodes, ...rest} = self;
  const cloned = structuredClone(rest) as Fragment;
  cloned.props = Props.cloneKnownProps(props);
  cloned.nodes = nodes;
  return cloned;
};

/**
 * Value
 */
export type ValueType =
  | string
  | bigint
  | number
  | boolean
  | null
  | undefined
  | symbol;

export const isValueType = (type: any): type is ValueType =>
  !type ||
  typeof type !== 'object';

export type Value =
  | string
  | bigint
  | number
  | true;

export const isValue = (self: Elem): self is Value =>
  typeof self !== 'object' ||
  self === null;

export const makeValue = (type: any): Value => {
  return type;
};

export const cloneValue = (self: Value) => structuredClone(self) as Value;
