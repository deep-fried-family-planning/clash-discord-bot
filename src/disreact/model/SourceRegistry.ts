import {DsxSettings} from '#src/disreact/runtime/DisReactConfig.ts'
import {FC} from '#src/disreact/model/entity/fc.ts'
import {Elem} from '#src/disreact/model/entity/elem.ts'
import {Root} from '#src/disreact/model/entity/root.ts'
import {Arr, Data, E, Hash} from '#src/disreact/re-exports.ts'
import type {Fibril} from './fibril/fibril'


export class SourceDefect extends Data.TaggedError('disreact/SourceDefect')<{
  message?: string
  why?    : string
  cause?  : Error
}> {}

const STORE = new Map<string, Root.Source>()

export class SourceRegistry extends E.Service<SourceRegistry>()('disreact/SourceRegistry', {
  accessors: true,

  effect: E.andThen(DsxSettings, (config) => {
    const sources = [
      ...config.sources.modal.map((src) => Root.make(Root.MODAL, src)),
      ...config.sources.public.map((src) => Root.make(Root.PUBLIC, src)),
      ...config.sources.ephemeral.map((src) => Root.make(Root.EPHEMERAL, src)),
    ]

    for (const src of sources) {
      // if (Î¾.has(src.id)) {
      //   return E.die(new SourceDefect({why: `Duplicate Source: ${src.id}`}));
      // }
      STORE.set(src.id, src)
    }

    const version = config.version
      ? `${config.version}`
      : Hash.array(Arr.map(sources, (src) => src.id))

    const checkout = (key: string | FC | Elem.Task, props: any = {}) => {
      const id = typeof key === 'string' ? key
        : Elem.isTask(key) ? FC.getSrcId(key.type)
          : FC.getSrcId(key)

      const src = STORE.get(id)

      if (!src) {
        return E.fail(new SourceDefect({message: `Unregistered Source: ${key}`}))
      }

      return E.succeed(Root.fromSource(src, props))
    }

    const fromHydrant = (hydrant: Fibril.Hydrant) => {
      const src = STORE.get(hydrant.id)

      if (!src) {
        return E.fail(new SourceDefect({message: `Unregistered Source: ${hydrant.id}`}))
      }

      return E.succeed(Root.fromSourceHydrant(src, hydrant))
    }

    return E.succeed({
      fromHydrant,
      checkout,
      version,
    })
  }),
}) {}
