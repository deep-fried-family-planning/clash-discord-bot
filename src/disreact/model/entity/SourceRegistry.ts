import {DsxSettings} from '#src/disreact/DisReactConfig.ts'
import {FC} from '#src/disreact/model/entity/comp/fc.ts'
import {Elem} from '#src/disreact/model/entity/elem/elem.ts'
import {Root} from '#src/disreact/model/entity/root.ts'
import {Arr, Data, E, Hash} from '#src/disreact/re-exports.ts'


export class SourceDefect extends Data.TaggedError('disreact/SourceDefect')<{
  message?: string
  why?    : string
  cause?  : Error
}> {}

const STORE = new Map<string, Root.Source>()

export class SourceRegistry extends E.Service<SourceRegistry>()('disreact/SourceRegistry', {
  accessors: true,

  effect: DsxSettings.use((config) => {
    const sources = [
      ...config.sources.modal.map((src) => Root.make(Root.MODAL, src)),
      ...config.sources.public.map((src) => Root.make(Root.PUBLIC, src)),
      ...config.sources.ephemeral.map((src) => Root.make(Root.EPHEMERAL, src)),
    ]

    for (const src of sources) {
      // if (Î¾.has(src.id)) {
      //   return E.die(new SourceDefect({why: `Duplicate Source: ${src.id}`}));
      // }
      STORE.set(src.id, src)
    }

    const checkout = (key: string | FC | Elem.Task, props?: any) => {
      const id = typeof key === 'string' ? key
        : Elem.isTask(key) ? FC.getSrcId(key.type)
          : FC.getSrcId(key)

      const src = STORE.get(id)

      if (!src) {
        return E.fail(new SourceDefect({message: `Unregistered Source: ${key}`}))
      }

      return E.succeed(Root.fromSource(src, props))
    }

    const version = config.version
      ? `${config.version}`
      : Hash.array(Arr.map(sources, (src) => src.id))

    return E.succeed({
      checkout,
      version,
    })
  }),
}) {}
