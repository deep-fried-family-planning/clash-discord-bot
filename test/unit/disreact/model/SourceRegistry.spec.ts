import { Elem } from '#src/disreact/model/entity/elem/elem.ts'
import {SourceRegistry} from '#src/disreact/model/entity/SourceRegistry.ts'
import {E} from '#src/internal/pure/effect.ts'
import {TestMessage} from 'test/components/test-message.tsx'
import {it} from 'test/components/TestRegistry.tsx'



describe('disreact/SourceRegistry', () => {
  it.effect('when hashing version', E.fn(function* () {
    const registry = yield* SourceRegistry

    expect(registry.version).toMatchInlineSnapshot(`981993584`)
  }))

  it.effect('when hashing version again', E.fn(function* () {
    const registry = yield* SourceRegistry

    expect(registry.version).toMatchInlineSnapshot(`981993584`)
  }))

  it.effect('when checking out a root', E.fn(function* () {
    const registry = yield* SourceRegistry
    const root     = yield* registry.checkout(TestMessage)
    Elem.linearizeElem(root.elem)
    expect(root).toMatchSnapshot()
  }))

  it.effect('when cloning a root', E.fn(function* () {
    const registry = yield* SourceRegistry
    const root     = yield* registry.checkout(TestMessage)
    const elem     = Elem.deepCloneElem(root.elem)
    Elem.linearizeElem(elem)
    expect(elem).toMatchSnapshot()
  }))
})
