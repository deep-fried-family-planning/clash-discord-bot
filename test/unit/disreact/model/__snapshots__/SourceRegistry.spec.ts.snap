// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`disreact/SourceRegistry > when checking out a root 1`] = `
{
  "_tag": "Ephemeral",
  "elem": {
    "id": "TestMessage",
    "idn": "TestMessage",
    "idx": undefined,
    "nodes": [],
    "props": {},
    "strand": {
      "pc": 0,
      "queue": [],
      "rc": 0,
      "saved": [],
      "stack": [],
    },
    "type": [Function],
  },
  "id": "TestMessage",
  "nexus": {
    "id": "TestMessage",
    "next": {
      "id": "TestMessage",
      "props": {},
    },
    "props": {},
    "queue": [],
    "root": [Circular],
    "strands": {
      "TestMessage": {
        "pc": 0,
        "queue": [],
        "rc": 0,
        "saved": [],
        "stack": [],
      },
    },
  },
}
`;

exports[`disreact/SourceRegistry > when cloning a root 1`] = `
{
  "id": "TestMessage",
  "idn": "TestMessage",
  "idx": undefined,
  "nodes": [],
  "props": {},
  "strand": {
    "pc": 0,
    "queue": [],
    "rc": 0,
    "saved": [],
    "stack": [],
  },
  "type": [Function],
}
`;
